File: Python/ScientificCalculator/functions/roots.py
"""
roots.py

Root functions.
"""

from decimal import Decimal
import math


def sqrt(x):
    return Decimal(str(x)).sqrt()


def cbrt(x):
    return Decimal(str(x)) ** (Decimal(1) / Decimal(3))


def root(n, x):
    """
    nth root of x.

    root(2, 25) -> 5
    root(3, 64) -> 4
    """

    return Decimal(str(x)) ** (Decimal(1) / Decimal(n))


def sqrtrem(x):
    """
    Integer square root and remainder.

    sqrtrem(26)
    -> 5 remainder 1
    """

    value = int(x)

    root = math.isqrt(value)

    remainder = value - (root * root)

    return f"{root} remainder {remainder}"

File: Java/ScientificCalculator/src/functions/Roots.java
package functions;

public class Roots {


    // ==========================
    // Square Root
    // ==========================

    public static double sqrt(
            double x
    ) {

        if (x < 0) {
            throw new ArithmeticException(
                "Cannot calculate square root of negative number"
            );
        }

        return Math.sqrt(x);
    }


    // ==========================
    // Cube Root
    // ==========================

    public static double cbrt(
            double x
    ) {

        return Math.cbrt(x);
    }


    // ==========================
    // nth Root
    // ==========================

    public static double root(
            double x,
            double n
    ) {

        if (n == 0) {
            throw new ArithmeticException(
                "Zeroth root is undefined"
            );
        }

        if (x < 0 && n % 2 == 0) {
            throw new ArithmeticException(
                "Even root of negative number is undefined"
            );
        }

        return Math.pow(
                x,
                1.0 / n
        );
    }


    // ==========================
    // Reciprocal
    // ==========================

    public static double reciprocal(
            double x
    ) {

        if (x == 0) {
            throw new ArithmeticException(
                "Division by zero"
            );
        }

        return 1.0 / x;
    }

}
